### ISSUE TITLE
Feature Request - As a theorist, I need the TheoryDataPoint context in ASSERT report hook

### ISSUE THREAD

**Issue by am11** [Snaipe/Criterion#40]
Hello,

In our current effort of switching from Ruby-based test runner to Criterion, I have encountered a scenario where we are alerting the observer about a TODO test, which unexpectedly passes:
https://github.com/sass/sass-spec/blob/6d3835/lib/sass_spec/test.rb#L26 &
https://github.com/sass/sass-spec/blob/6d3835/lib/sass_spec/test.rb#L37

In order to replicate that effect with Criterion, we have experimented [with ASSERT report hook](https://github.com/am11/libsass/blob/2435ace5e/test/spec_theories.c#L142) which although emits the status of assertion `test->passed`, but does not provide context for Theory's arguments or expected/actual result. Analogous to `struct criterion_theory_stats` for theory, it would be of great help if the existing `struct criterion_assert_stats` or a new `struct criterion_theory_assert_stats` let us exercise:

```c
// Supposedly, it requires a separate `THEORY_ASSERT`
ReportHook(THEORY_ASSERT)(struct criterion_theory_assert_stats *test) {
  if(test->passed && should_test_todo && strstr(test->formatted_args, "-todo-")) {
    // alert unexpectedness 
  }
}
```
---
Nitpick:
On a related note, it seems like `test->message` always contains the error string, even when `test->passed` is `true`.

**Comment by Snaipe** [comment]
The issue with exposing the data points to a generic report hook is that it would effectively be quite error prone; you'd have a (size_t, void*) tuple that you would need to correctly manage on the report hook end, because the parameter tuple is theory-specific while the hook runs at a global scope.

From what I gather in what you want to do, is that you would like to expect a test to fail. That feature is currently lacking, but can be added without any trouble.

Effectively, you'd then have two theories:

```c
void test_for_input(const char *input_path);

Theory((const char *input_path), my_suite, todo_tests, .expect_fail = 1) {
    cr_assume(strstr(input_path, "-todo-"));
    test_for_input(input_path);
}

Theory((const char *input_path), my_suite, non_todo_tests) {
    cr_assume(!strstr(input_path, "-todo-"));
    test_for_input(input_path);
}
```

(assuming a common dataset).

Would that be better ?

**Comment by am11** [comment]
I see this is an interesting approach. However, I think I have missed some info in the first post:

* I have elected to use an [evn var knob](https://github.com/am11/libsass/blob/2435ace5ec8/test/spec_theories.c#L128), which controls the TODO test behavior (LIBSASS_TEST_TODO==1 => test todos, else don't).
* When the TODO tests are enabled, we expect it to fail the test, cross the fingers and hope for surprises, if some TODO test magically start passing! :)
* Normally the owners run the TODO tests locally when refactoring/shuffling the directory structure.
* Therefore, we only need to *alert* the tester if there transpires a (surprise) progression

(@xzyfer, please correct me, if my understanding is wrong)

At this point, the intention is to make it drop-in replacement, as close to Ruby version as possible. Later, we will diverge and incorporate sanity to it. :smile: 

BTW, currently I am generating only one array of paths. Splitting it into two will not be an issue.

**Comment by Snaipe** [comment]
Aaah, I see.

Reading the ruby source, it seems that the best way to transcribe the `raise` is:

```c
if (should_test_todo && strstr(input_path, "-todo-") && !strcmp(expected, clean_output))
    cr_assert_fail("%s passed a test we expected it to fail", input_path);
```

from within the theory itself.

**Comment by xzyfer** [comment]
@am11 the current LibSass work flow is.

#### Locally

- run active specs during development (to catch regressions)
  - optionally enable specific TODO specs (to make sure your patch is working)
- occasionally run spec suite in "unexpected mode" which errors when TODO specs surprisingly pass

#### CI

- run active specs from spec suite HEAD
  - optionally supply a spec suite branch to test against (generally with TODO specs for your patch enabled to ensure your patch works)

**Comment by am11** [comment]
@Snaipe, @xzyfer, thanks for the inputs!

@Snaipe, reading the Ruby source again, your comment and @xzyfer's, I think I have the knob in reverse order. It should be `LIBSASS_SKIP_TODO`, then we should have a `LIBSASS_UNEXPECTED_MODE` etc. 
With your example, I think all required varieties can be handled within the `Theory`. Evidently, there is more than one way to skin this cat! :joy_cat: 
I would just try to avoid two comparisons; one made by `strcmp` and other `cr_assert_str_eq` by caching the `strcmp`'s result in a `bool` and use `cr_assert` & `cr_assert_fail` accordingly. I think if we get the `TheoryDataPoint` context in reporter, the only difference would be that we will move that branching logic in reporter's body instead of `Theory` itself (but in the latter assumed case, I am not sure if throwing an assertion failure within a reporter which was triggered by another assertion would be a good idea).

**Comment by Snaipe** [comment]
Well regarding the double comparison, if it ever becomes a performance problem (which I expect won't, unless you compare 2GB strings), you can still use the more "manual" method:

```c
int same_output = !strcmp(expected, actual);

cr_assert(same_output, ...);
if (... && same_output)
    cr_assert_fail(...);
```

**Comment by Snaipe** [comment]
@am11 btw, there are some new assertion macros to test more easily file contents:

```c
FILE* reference_file = fopen(ref_file_path, "r")
cr_assert_file_contents_eq_str(reference_file, actual_output);
fclose(reference_file);
```

**Comment by am11** [comment]
Fantastic, I will try it out. But I am afraid we need to sanitize expected string as well before making the comparison. Would be nice (and lame from Criterion's reusability perspective) if could pass a sanitize function reference to this assertion! :D
For that matter, I am looking into PCRE and SLRE options, as you suggested. Although POSIX `<regex.h>` is not available in VS, but C++11 `<regex>` is. But I am trying not to mix C and C++ at this point and go with pure C. Hopefully SLRE will do the job. :)

**Comment by am11** [comment]
The SLRE-based `sanitize_str` is incorporated with: https://github.com/am11/libsass/commit/135a8ed40492ebe4ad47377ab08c98c6751bc078 :)

SLRE apparently does not have support for multiline mode, so I had to sprinkle some `(\r?\n)*` to mimic this ruby regex https://github.com/sass/sass-spec/blob/6d38353/lib/sass_spec/test_case.rb#L69-L75. Also the capture replace does not have ability to [implode nested capture in substitution](https://github.com/sass/sass-spec/blob/6d38353/lib/sass_spec/test_case.rb#L71), so I had to split those rules, although I made an alternative version of replace to allow multiple captures(`slre_replace_alternative_dynamic_capture_version`, which their original unit_test version does not) but it needs some work to make use of all the captured groups. Finally, in Ruby `\s` implies `[<whitespace>\t\r\n\f]`, while in SLRE, it only is for some special whitespace (but apparently not ASCII-32), so had to split them as well. :)

All the ~1240 fixtures pass with `SASS_STYLE_NESTED` (now need to repeat the test for `COMPAT`, `COMPRESSED` and `EXPANDED` styles).

I will incorporate the `TODO` workflow and aforementioned improvements next.

@Snaipe, at present -- with the suggested way, it is testing all the assets in one Theory. I am trying to figure out a way to segregate it with one input x 4 styled_outputs per theory. For example, by creating `TheoryDataPoints` in a loop on `list_directories()` output and save the reference to `struct criterion_datapoints` in an array outside of `generate_datapoints`. I think in future, if we are able to run the tests in parallel with Criterion (https://github.com/Snaipe/Criterion/issues/39), separate theories would make sense. Also, it is required to get the meaningful stats: ~1240 x 4 tests in 1 Theory vs ~1240 x 4 tests in 1240 theories.
Is it worth the effort? Or should we wait for some syntactic sugary macro for this? =)

**Comment by Snaipe** [comment]
I'm not quite sure to understand what you're saying.

From what I gather, you're saying that the theories should each count as a full blown test, correct ?

Currently, theories are implemented a bit differently, as they all run in one test; this means that a theory is actualy a promise of an unified behaviour over a cartesian product of multiple datasets rather than an iteration over a dataset, and should be treated as a full blown test.

I gave the issue some thought since you first opened it, and I think the reason why the usage of theories for the purpose of genericity is a bit syntactically unfortunate is probably because theories are not exactly meant to run one particular test with a particular parameter.
The ideal solution here would be to implement parameterized tests, that would iterate over a user-generated dataset rather than the cartesian product of many datasets.

**Comment by am11** [comment]
> The ideal solution here would be to implement parameterized tests, that would iterate over a user-generated dataset rather than the cartesian product of many datasets.

Yup this will eventually benefit us in getting correct total count in test results. Other than that, your [placeholder workaround](https://github.com/am11/libsass/blob/64267fce0df339c876cb876d17cf58e9c65d429e/test/spec_theories.c#L144) for theories is doing a brilliant job. :heavy_check_mark:

**Comment by am11** [comment]
The unexpect_pass, skip_todo and the formatted failure messages based on the environment variables we have discussed here are incorporated with: https://github.com/am11/libsass/commit/bd1109893ed13d3257049f4ca3e04831a55e4c23.

For the user-defined dataset, I have opened issue #43.

@Snaipe, should this be closed? The <kbd>Feature</kbd> label was added to this issue, but our concern was addressed by using a different approach.
